src/canonicalization.jl

```diff
@@ -0,0 +1,170 @@
+        else 
+            # add a CNOT, and measure to the front. New ops becomes -> [CNOT, MEAS, ...]
+            # using pushfirst! so this is done in opposite order.
+            # control will be a purified pair, and target is this pair
+            pushfirst!(ops,rand(BellMeasure,pair))
+            pushfirst!(ops,rand(CNOTPerm,rand(1:num_purified),pair))
+        end
+    end
+    
+    return ops
+end
+
+"""
+    cleanup_measurements_on_top_qubits!(ops,num_purified)
+
+    Checks for and removes any measurements on purified pairs
+"""
+function cleanup_measurements_on_top_qubits!(ops,num_purified)
+    return filter!(op-> 
+        !(  # remove it if,
+            typeof(op) <: AbstractMeasurement # it is a measurement and,
+            && op.sidx <= num_purified # it measures a purified pair
+        ), ops)
+end
+
+"""
+    cleanup_nonmeasurement_last_steps!(ops,num_pairs,num_purified)
+
+    If there is a non-measurement in the last step of a non-purified pair, add measurements.
+"""
+function cleanup_nonmeasurement_last_steps!(ops,num_pairs,num_purified)
+     if num_pairs <= num_purified 
+        @warn "Not enough pairs - skipping canonicalization step cleanup_nonmeasurement_last_steps!"
+        return ops
+    end
+
+    non_pure_last_steps = zeros(Int64,num_pairs - num_purified) # array id, plus the num_purified, is the non_pure pair number. The contents of the array 
+    notDone = true 
+    while notDone
+        # get all of the last steps
+        for (op_index,op) in enumerate(ops)
+            qubits = affectedqubits(op)
+            for qubit in qubits
+                if qubit > num_purified
+                    # this is a non purified pair, mark a new last step
+                    non_pure_last_steps[qubit - num_purified] = op_index
+                end
+            end
+        end
+
+        # check if any of the last ops are not measurements, and ignore zeros (no ops on that qubit)
+        qubits_who_need_measurements = reduce((arr, (qubit, op_index)) ->
+                (op_index == 0 # filter out zeros
+                || typeof(ops[op_index]) <: AbstractMeasurement) ? arr : # and filter out measurments
+                push!(arr,qubit+num_purified), # otherwise, keep it and add 'num_purified' so the qubit index is correct
+            enumerate(non_pure_last_steps);init=[])
+        
+        # all of the marked ops are non_last measurements, so add measures to them 
+        for qubit in qubits_who_need_measurements
+            push!(ops,rand(BellMeasure,qubit))
+        end
+        
+        # check if done
+        if length(qubits_who_need_measurements) == 0
+            notDone = false
+        end
+    end
+    return ops
+end
+
+"""
+    canonicalize_cleanup!(pop::Population,num_pairs,num_purified)
+
+    Apply all of the canonicalization cleanup methods to a population with tmap.
+"""
+function canonicalize_cleanup!(pop::Population,num_pairs,num_purified)
```

Comment:
maybe add a flag, `canon_type`

- `:all` just run all of your canonicalizations
- `:safe` run only the safe canonicalizations and drop the ones that are matched by the unsafe ones
- maybe others

or maybe sturcture it differently if that makes the code more pleasant

Two use cases:
- towards the start of optimization, used for mangling lousy circuits
- at the end for just removing bad circuits